File: Recursion/combination_sum.cpp
/*
https://leetcode.com/problems/combination-sum/description/
39. Combination Sum

Given an array of distinct integers candidates and a target integer target,
return a list of all unique combinations of candidates where the chosen numbers
sum to target. You may return the combinations in any order.

The same number may be chosen from candidates an unlimited number of times.
Two combinations are unique if the
frequency of at least one of the chosen numbers is different.

The test cases are generated such that the number of unique combinations
that sum up to target is less than 150 combinations for the given input.



Example 1:

Input: candidates = [2,3,6,7], target = 7
Output: [[2,2,3],[7]]
Explanation:
2 and 3 are candidates, and 2 + 2 + 3 = 7. Note that 2 can be used multiple
times. 7 is a candidate, and 7 = 7. These are the only two combinations. Example
2:

Input: candidates = [2,3,5], target = 8
Output: [[2,2,2,2],[2,3,3],[3,5]]
Example 3:

Input: candidates = [2], target = 1
Output: []


Constraints:

1 <= candidates.length <= 30
2 <= candidates[i] <= 40
All elements of candidates are distinct.
1 <= target <= 40
*/

#include <iostream>
#include <vector>

using namespace std;

class Solution
{
private:
  void dfs(const vector<int>& candidates, int index, vector<int>& curr,
           int target, vector<vector<int>>& result)
  {
    if (target == 0)
    {
      result.push_back(curr);
      return;
    }

    if (target < 0 || index == candidates.size())
    {
      return;
    }

    // Include the current candidate
    // curr.push_back(candidates[index]);
    // dfs(candidates, index, curr, target - candidates[index], result);
    // curr.pop_back();

    // Exclude the current candidate
    // dfs(candidates, index + 1, curr, target, result);

    // Iterate through candidates starting from the current index
    for (int i = index; i < candidates.size(); i++)
    {
      curr.push_back(candidates[i]);
      dfs(candidates, i, curr, target - candidates[i], result);
      curr.pop_back();
    }
  }

public:
  vector<vector<int>> combinationSum(vector<int>& candidates, int target)
  {
    vector<vector<int>> result;
    vector<int> curr;
    dfs(candidates, 0, curr, target, result);
    return result;
  }
};

int main()
{
  cout << "Enter the number of elements in the candidates array: ";
  int n;
  cin >> n;

  cout << "Enter the elements of the candidates array: ";
  vector<int> candidates(n, 0);
  for (int i = 0; i < n; i++)
  {
    cin >> candidates[i];
  }

  cout << "Enter the target sum: ";
  int target;
  cin >> target;

  Solution sol;
  vector<vector<int>> result = sol.combinationSum(candidates, target);

  cout << "\nOutput combinations:\n[";
  for (size_t i = 0; i < result.size(); ++i)
  {
    cout << "[";
    for (size_t j = 0; j < result[i].size(); ++j)
    {
      cout << result[i][j];
      if (j < result[i].size() - 1)
        cout << ",";
    }
    cout << "]";
    if (i < result.size() - 1)
      cout << ",";
  }
  cout << "]" << endl;

  return 0;
}

File: recursion/CombinationSum.java
/*
 * https://takeuforward.org/data-structure/combination-sum-1/
 * https://youtu.be/OyZFFqQtu98
 * 
 * https://leetcode.com/problems/combination-sum/description/
*/

package recursion;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class CombinationSum {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] nums = new int[n];
        for (int i = 0; i < n; i++) {
            nums[i] = sc.nextInt();
        }

        int target = sc.nextInt();

        CombinationSum sol = new CombinationSum();
        System.out.println(sol.combinationSum(nums, target));

        sc.close();
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> curr = new ArrayList<>();
        List<List<Integer>> res = new ArrayList<>();

        backtrack(candidates, candidates.length - 1, target, curr, res);

        backtrack2(candidates, 0, target, curr, res);

        return res;
    }

    /**
     * Using pick and non pick technique
     */
    private void backtrack(int[] nums, int n, int target, List<Integer> curr, List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<>(curr));
            return;
        }

        if (target < 0 || n < 0)
            return;

        curr.add(nums[n]);
        backtrack(nums, n, target - nums[n], curr, res);
        curr.remove(curr.size() - 1);

        backtrack(nums, n - 1, target, curr, res);

        // TODO: Attach recursion tree
    }

    /**
     * Using looping through choices.
     * Generate all combinations (with repetition) of nums[index…] that sum to
     * target.
     */
    private void backtrack2(int[] nums, int index, int target, List<Integer> curr, List<List<Integer>> res) {
        if (target == 0) {
            res.add(new ArrayList<>(curr));
            return;
        }

        if (target < 0 || index == nums.length) {
            return;
        }

        for (int i = index; i < nums.length; i++) {
            curr.add(nums[i]);
            backtrack2(nums, i, target - nums[i], curr, res);
            curr.remove(curr.size() - 1);
        }

        // TODO: Attach recursion tree
    }
}

// Recursive Time Complexity: Much more than O(2^N) , beacuse at each step we
// have the choice of picking the current element again and again
// pick or not pick

// Recursive Space Complexity: O(N)

// Overall Space Complexity: O(N * 2^N + N), more than that
